crates/unftp-sbe-gcs/src/gcs_client.rs

```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    async fn http_raw<B>(&self, method: Method, uri: Uri, body: B, headers: &[(&str, &str)]) -> Result<Response<Body>, Error>
+    where
+        B: Into<Body>,
+    {
+        let token = self.tokens.token().await?;
+        let mut request = Request::builder().uri(uri).header(header::AUTHORIZATION, format!("Bearer {}", token));
+
+        for (hk, hv) in headers.as_ref() {
+            request = request.header(*hk, *hv);
+        }
+
+        let request = request
+            .method(method)
+            .body(body.into())
+            .map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))?;
+
+        let response = self
+            .http
+            .request(request)
+            .await
+            .map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))?;
+
+        if !response.status().is_success() {
+            let err_kind = match response.status().as_u16() {
+                404 => ErrorKind::PermanentFileNotAvailable,
+                401 | 403 => ErrorKind::PermissionDenied,
+                429 => ErrorKind::TransientFileNotAvailable,
+                _ => ErrorKind::LocalError,
+            };
+
+            return Err(Error::from(err_kind));
+        }
+
+        Ok(response)
+    }
+
+    async fn http_delete_raw(&self, uri: Uri) -> Result<Response<Body>, Error> {
+        self.http_raw(Method::DELETE, uri, Body::empty(), &[]).await
+    }
+
+    async fn http_get_raw(&self, uri: Uri, headers: &[(&str, &str)]) -> Result<Response<Body>, Error> {
+        self.http_raw(Method::GET, uri, Body::empty(), headers).await
+    }
+
+    async fn http_post_raw(&self, uri: Uri, body: Body, headers: &[(&str, &str)]) -> Result<Response<Body>, Error> {
+        self.http_raw(Method::POST, uri, body, headers).await
+    }
+
+    async fn http_post<T>(&self, uri: Uri, body: Body, headers: &[(&str, &str)]) -> Result<T, Error>
+    where
+        T: DeserializeOwned,
+    {
+        let response = self.http_post_raw(uri, body, headers).await?;
+
+        deserialize(response).await
+    }
+
+    async fn http_get<T>(&self, uri: Uri) -> Result<T, Error>
+    where
+        T: DeserializeOwned,
+    {
+        let response = self.http_get_raw(uri, &[]).await?;
+
+        deserialize(response).await
+    }
+}
+
+enum TrailingSlash {
+    Trim,
+    Ensure,
+    AsIs,
+}
+
+async fn deserialize<T>(response: Response<Body>) -> Result<T, Error>
+where
+    T: DeserializeOwned,
+{
+    let body = hyper::body::aggregate(response)
+        .await
+        .map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))?;
+
+    serde_json::from_reader(body.reader()).map_err(|e| Error::new(ErrorKind::PermanentFileNotAvailable, e))
+}
+
+fn make_uri(path_and_query: String) -> Result<Uri, Error> {
+    Uri::from_maybe_shared(path_and_query).map_err(|_| Error::from(ErrorKind::FileNameNotAllowedError))
+}
+
+#[derive(Clone, Debug)]
+struct TokenManager {
```

Comment:
I hate the word manager, it never indicates what something is actually doing. If we can't find a better name then lets at least add a description indicating its intent.

Reply:
Yeah, me too. I took this name from another Google APIs client. I renamed it to TokenSource. Let me know whether that's better or not

Reply:
Much better already, thanks